File: Java/src/htmltest/base/BaseTest.java
package htmltest.base;

import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

import org.junit.Test;

import htmlwalker.platform.WalkerPlatform;

public abstract class BaseTest
{
    protected class Options
    {
        public final WalkerPlatform.IOptions documentOptions;

        public boolean formatted = true;

        public Options(WalkerPlatform platform)
        {
            documentOptions = platform.newDocumentOptions();
        }
    }

    private final File tstFolder;
	private final File outFolder;
	
	protected abstract String testFolder();
	protected abstract String outFolder();
	
	protected abstract WalkerPlatform platform();
	
    protected abstract void createExpected(File expected, File testInput);
	
	public BaseTest()
	{
		this.tstFolder = new File(testFolder());
		this.outFolder = new File(outFolder());
	}
	
	protected BaseTest(String subFolder)
	{
		if (null == subFolder)
		{
			this.tstFolder = new File(testFolder());
			this.outFolder = new File(outFolder());
		}
		else
		{
			this.tstFolder = new File(testFolder() + File.separator + subFolder);
			this.outFolder = new File(outFolder() + File.separator + subFolder);
		}
	}

	@Test
	public void t0confirmTestFolder()
	{
		try
		{
			assertTrue(this.tstFolder.getCanonicalPath() + " does not exists", this.tstFolder.getCanonicalFile().exists());
		}
		catch (IOException e)
		{
			e.printStackTrace();
			fail(e.getMessage());
		}
	}

	@Test
	public void t0confirmOutFolder()
	{
		try
		{
			assertTrue(this.outFolder.getCanonicalPath() + " does not exists", this.tstFolder.getCanonicalFile().exists());
		}
		catch (IOException e)
		{
			e.printStackTrace();
			fail(e.getMessage());
		}
	}

	protected String getTestPath(String filename)
	{
		try
		{
			return this.tstFolder.getCanonicalPath() + File.separator + filename;
		}
		catch (IOException e)
		{
			e.printStackTrace();
			return null;
		}
	}

	protected String getOutPath(String filename)
	{
		try
		{
			return this.outFolder.getCanonicalPath() + File.separator + filename;
		}
		catch (IOException e)
		{
			e.printStackTrace();
			return null;
		}
	}

    protected File getTestInput(String filename)
    {
	    return new File(getTestPath(filename));
    }

    protected File getTestOutput(String filename)
    {
	    return new File(getOutPath(filename));
    }    

    protected File getTestExpected(String filename, File testInput)
    {
	    File expected = new File(getOutPath(filename));
	    if (!expected.exists())
		     createExpected(expected, testInput);
	    return expected;
    }

    private String readNextLine(BufferedReader reader, boolean ignoreWhitespace) throws IOException
	{
        String line = reader.readLine();
        if (ignoreWhitespace)
        {
        	while (null != line && line.trim().length() == 0)
        		line = reader.readLine();
        }
	    return line;
	}

	protected boolean compareLines(String expected, String output, boolean ignoreWhitespace)
	{
	    boolean match = true;
		
	    String eLine = expected.trim();
	    String oLine = output.trim();
	
	    int ePos = 0;
	    int oPos = 0;
	
	    while (match && ePos < eLine.length() && oPos < oLine.length())
	    {
	    	if (ignoreWhitespace)
	    	{
	    		while (Character.isWhitespace(eLine.charAt(ePos)) && ePos < eLine.length())
	    			ePos++;
		
	    		while (Character.isWhitespace(oLine.charAt(oPos)) && oPos < oLine.length())
	    			oPos++;
	    	}
		
		    match = eLine.charAt(ePos) == oLine.charAt(oPos);
		
		    ePos++;
		    oPos++;
	    }
	
	    if (match)
		    match = ePos == eLine.length() && oPos == oLine.length();
	
	    return match;
	}

    protected void compareToExpected(File testOutput, File testInput, String expectedFile, boolean ignoreWhitespace)
    {
        File testExpected = getTestExpected(expectedFile, testInput);
        compareToExpected(testOutput, testExpected, ignoreWhitespace);
    }

    protected void compareToExpected(File testOutput, File testExpected, boolean ignoreWhitespace)
	{
	    String failure = null;
	    
	    try
	    {
	        try (BufferedReader output = new BufferedReader(new FileReader(testOutput.getPath())))
	        {
	            try (BufferedReader expected = new BufferedReader(new FileReader(testExpected.getPath())))
	            {
	            	// TODO: Confirm not empty in sharp
	                // First line is declarations
	                String outputLine = output.readLine();
	                
	                while (outputLine != null && !outputLine.contains("<html"))
	                    outputLine = output.readLine();
	                
	                if (outputLine != null)
	                	outputLine = outputLine.substring(outputLine.indexOf("<html"));
	
	                String expectedLine = expected.readLine();
	                while (expectedLine != null && !expectedLine.contains("<html"))
	                    expectedLine = expected.readLine();
	                
	                if (expectedLine != null)
	                	expectedLine = expectedLine.substring(expectedLine.indexOf("<html"));

	                if (null != outputLine && !outputLine.isEmpty() && null != expectedLine && !expectedLine.isEmpty())
	                {
	                    outputLine = readNextLine(output, ignoreWhitespace);
	                    expectedLine = readNextLine(expected, ignoreWhitespace);
	                    while (null != outputLine && null != expectedLine && null == failure)
	                    {
	                        if (!compareLines(expectedLine, outputLine, ignoreWhitespace))
	                            failure = "Expected - [" + expectedLine + "] vs Out - [" + outputLine + "]";
	                        outputLine = readNextLine(output, ignoreWhitespace);
	                        expectedLine = readNextLine(expected, ignoreWhitespace);
	                    }
	                }
	                else
	                    failure = "Nothing to test";
	
	                if ((null != outputLine || null != expectedLine) && null == failure)
	                    failure = "One file is larger";
	            }
	            catch(Exception e)
	            {
	            	throw e;
	            }
	        }
            catch(Exception e)
            {
            	throw e;
            }
	    }
	    catch(Exception e)
	    {
	    	e.printStackTrace();
	    	failure = e.getMessage();
	    }

	    if (null != failure)
		    fail(failure);
	}

    protected void compareToExpected(File testOutput, String expected, boolean ignoreWhitespace)
    {
        String output = "";
        try (BufferedReader reader = new BufferedReader(new FileReader(testOutput.getPath())))
        {
        	String outputLine = readNextLine(reader, ignoreWhitespace);
            while (null != outputLine)
            {
                output += outputLine + System.lineSeparator();
                outputLine = readNextLine(reader, ignoreWhitespace);
            }
        }
        catch (IOException e)
        {
			e.printStackTrace();
			fail(e.getMessage());
		}

        if (!expected.equals(output))
            fail("Expected - [" + expected + "] vs Out - [" + output + "]");
    }
}

File: CSharp/HtmlTest/Base/BaseTest.cs
using HtmlWalker.Platform;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.IO;

namespace HtmlTest.Base
{
    [TestClass]
    public abstract class BaseTest
    {
        public class Options
        {
            public readonly WalkerPlatform.IOptions DocumentOptions;

            public bool Formatted = true;
            public bool AutoCreate = true;

            public Options(WalkerPlatform platform)
            {
                DocumentOptions = platform.NewDocumentOptions();
            }
        }

        public void PrintStackTrace(Exception e)
        {
            Console.WriteLine(e.Message);
            Console.WriteLine(e.StackTrace);
        }

        protected static string TestFolder => "../../../../Test";
        protected static string OutFolder => "../../../../Test";

        protected readonly DirectoryInfo InputDirectory;
        protected readonly DirectoryInfo OutputDirectory;

        protected abstract WalkerPlatform Platform { get; }

        protected abstract void CreateExpected(FileInfo expected, FileInfo testInput, Options options = null);

        public BaseTest()
        {
            InputDirectory = new DirectoryInfo(TestFolder);
            OutputDirectory = new DirectoryInfo(OutFolder);
        }

        protected BaseTest(string subFolder)
        {
            if (null == subFolder)
            {
                InputDirectory = new DirectoryInfo(TestFolder);
                OutputDirectory = new DirectoryInfo(OutFolder);
            }
            else
            {
                InputDirectory = new DirectoryInfo(TestFolder + Path.DirectorySeparatorChar + subFolder);
                OutputDirectory = new DirectoryInfo(OutFolder + Path.DirectorySeparatorChar + subFolder);
            }
        }

        [TestMethod]
        public virtual void TConfirmTestFolder()
        {
            try
            {
                Assert.IsTrue(InputDirectory.Exists, InputDirectory.FullName + " does not exist");
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        [TestMethod]
        public virtual void TConfirmOutFolder()
        {
            try
            {
                Assert.IsTrue(OutputDirectory.Exists, OutputDirectory.FullName + " does not exist");
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }
        }

        protected virtual string GetTestPath(string filename)
        {
            string path = null;

            try
            {
                path = InputDirectory.FullName + Path.DirectorySeparatorChar + filename;
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }

            return path;
        }

        protected virtual string GetOutPath(string filename)
        {
            string path = null;

            try
            {
                path = OutputDirectory.FullName + Path.DirectorySeparatorChar + filename;
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                Assert.Fail(e.Message);
            }

            return path;
        }

        public FileInfo GetTestInput(string filename) => new FileInfo(GetTestPath(filename));

        public FileInfo GetTestOutput(string filename) => new FileInfo(GetOutPath(filename));

        public FileInfo GetTestExpected(string filename, FileInfo testInput, Options options = null)
        {
            Assert.IsTrue(testInput.Exists, testInput.FullName + " did not exists.");

            var outPath = GetOutPath(filename);
            var expected = new FileInfo(outPath);
            if (!expected.Exists)
            {
                CreateExpected(expected, testInput, options);
                Assert.IsTrue(File.Exists(outPath), outPath + " did not exists and could not be created.");
                if (!(options?.AutoCreate ?? true))
                    Assert.Inconclusive(filename + " had to be created. Rerun test.");
            }
            return expected;
        }

        private string ReadNextLine(TextReader reader, bool ignoreWhitespace)
        {
            string line = reader.ReadLine();
            if (ignoreWhitespace)
            {
                while (null != line && line.Trim().Length == 0)
                    line = reader.ReadLine();
            }
            return line;
        }

        protected bool CompareLines(string expected, string output, bool ignoreWhitespace = true)
        {
            bool match = true;

            string eLine = expected;
            string oLine = output;

            if (ignoreWhitespace)
            {
                eLine = eLine.Trim();
                oLine = oLine.Trim();
            }

            int ePos = 0;
            int oPos = 0;

            while (match && ePos < eLine.Length && oPos < oLine.Length)
            {
                if (ignoreWhitespace)
                {
                    while (Char.IsWhiteSpace(eLine[ePos]) && ePos < eLine.Length)
                        ePos++;

                    while (Char.IsWhiteSpace(oLine[oPos]) && oPos < oLine.Length)
                        oPos++;
                }

                match = eLine[ePos] == oLine[oPos];

                ePos++;
                oPos++;
            }

            if (match)
                match = ePos == eLine.Length && oPos == oLine.Length;

            return match;
        }

        protected void CompareToExpected(
            FileInfo testOutput, FileInfo testInput, string expectedFile, bool ignoreWhitespace = true
        ) {
            FileInfo testExpected = GetTestExpected(expectedFile, testInput);
            CompareToExpected(testOutput, testExpected, ignoreWhitespace);
        }

        protected virtual void CompareToExpected(
            FileInfo testOutput, FileInfo testExpected, bool ignoreWhitespace = true
        ) {
            string failure = null;
            try
            {
                using TextReader output = new StreamReader(testOutput.Open(FileMode.Open, FileAccess.Read, FileShare.Read));
                using (TextReader expected = new StreamReader(testExpected.Open(FileMode.Open, FileAccess.Read, FileShare.Read)))
                {
                    string outputLine = output.ReadLine();
                    if (testOutput.Extension == ".html" || testOutput.Extension == ".htm")
                    {
                        // First line is declarations
                        while (!outputLine.Contains("<html"))
                            outputLine = output.ReadLine();
                        outputLine = outputLine.Substring(outputLine.IndexOf("<html"));
                    }

                    string expectedLine = expected.ReadLine();
                    if (testExpected.Extension == ".html" || testExpected.Extension == ".htm")
                    {
                        // First line is declarations
                        while (!expectedLine.Contains("<html"))
                            expectedLine = expected.ReadLine();
                        expectedLine = expectedLine.Substring(outputLine.IndexOf("<html"));
                    }

                    if (null != outputLine && null != expectedLine)
                    {
                        while (null != outputLine && null != expectedLine && null == failure)
                        {
                            if (!CompareLines(expectedLine, outputLine, ignoreWhitespace))
                                failure = "Expected - [" + expectedLine + "] vs Out - [" + outputLine + "]";
                            outputLine = ReadNextLine(output, ignoreWhitespace);
                            expectedLine = ReadNextLine(expected, ignoreWhitespace);
                        }
                    }
                    else
                    {
                        failure = "Nothing to test";
                    }

                    if ((null != outputLine || null != expectedLine) && null == failure)
                        failure = "One file is larger";
                }
            }
            catch (Exception e)
            {
                PrintStackTrace(e);
                failure = e.Message;
            }

            if (null != failure)
                Assert.Fail(Platform.Name + " : " + failure);
        }

        protected void CompareToExpected(FileInfo testOutput, string expected)
        {
            string output = "";
            using (var reader = new StreamReader(testOutput.OpenRead()))
            {
                while (!reader.EndOfStream)
                    output += reader.ReadLine();
            }

            if (expected != output)
                Assert.Fail("Expected - [" + expected + "] vs Out - [" + output + "]");
        }
    }
}
